### DISCUSSION TITLE
How to use plugins/myPlugin.server.ts in server/api

### DISCUSSION THREAD

**Discussion by Fire7** [nuxt/nuxt#20114] [Questions] [answered]
Hello, 

I'm currently working on analytics module for Nuxt 3 and it should use different libraries for browser and node environment. 
I created `plugin.client.ts` and `plugin.server.ts` in `runtime` directory of the module, but I totally don't understand how to use server plugin in `server/*` directory inside the `defineEventHandler()` function. 

Thanks for the answers!

**Comment by steadygaze** [comment, marked as answer]
I've also been trying to figure this out, so I have a place to put server-side initialization that I only want to happen once. My understanding is that anything placed under `server/plugins` is a "Nitro plugin" as documented here: https://nitro.unjs.io/guide/plugins. It can access `nitroApp` as a parameter, and the `useRuntimeConfig()` composable. However, I don't see how we can export a function or some other value to be used later in a `server/api` route/`defineEventHandler`. E.g. just adding it as a property on `nitroApp` feels hacky and gives a Typescript error. Nuxt plugins (https://nuxt.com/docs/guide/directory-structure/plugins#automatically-providing-helpers) have a nice `provide` key to do this, but they're client-side.

The workaround I settled on is to wrap a function in lodash.once so it's always [lodash.once](https://lodash.com/docs/#once), export it from a module, and depend on Node.js module caching ([1](https://medium.com/@mackplevine/node-js-module-caching-10c4e395423e), [2](https://nodejs.org/api/modules.html)) to ensure it's only called once. There also might need to be some manual waiting on setTimeout if there's a chance your code will try to get the result before it's done. I was hoping not to do this because it's hacky but it seems to be working well enough.

**Reply by Fire7** [reply]
Thanks for the answer. 
Yes, I came to absolutely the same answer by plunging deeper into the question. The only thing I can add than `plugins/`, `composables/` are available only in `vue` context, so they are not available in `server/` context where `nitro` rules the show.
And I hope `provide` feature will be available for plugins in nitro soon too.

**Comment by jirehnimes** [comment]
This is not an answer. I created a discussion of the same inquiry in Nitro but my usage actually is for my Nuxt app as well. The documentation of `server/plugins` both in Nitro and Nuxt lacks details on how can it be used.

https://github.com/unjs/nitro/discussions/1221

**Reply by steadygaze** [reply]
I added the workaround/hack I settled on to my answer. But support from the Nitro side would still be best.

**Reply by jirehnimes** [reply]
@steadygaze Yeah totally agree especially for Typescript! 🙂
